lifemonitor/auth/oauth2/server/models.py

```diff
@@ -0,0 +1,180 @@
+        self.register_grant(PasswordGrant)
+        # register it to grant endpoint
+        self.register_grant(ClientCredentialsGrant)
+        # register it to grant endpoint
+        self.register_grant(RefreshTokenGrant)
+
+    @staticmethod
+    def create_client(user,
+                      client_name, client_uri,
+                      grant_type, response_type, scope,
+                      redirect_uri,
+                      token_endpoint_auth_method=None):
+        client_id = gen_salt(24)
+        client_id_issued_at = int(time.time())
+        client = Client(
+            client_id=client_id,
+            client_id_issued_at=client_id_issued_at,
+            user_id=user.id,
+        )
+
+        client_metadata = {
+            "client_name": client_name,
+            "client_uri": client_uri,
+            "grant_types": grant_type,
+            "redirect_uris": redirect_uri,
+            "response_types": response_type,
+            "scope": scope,
+            "token_endpoint_auth_method": token_endpoint_auth_method
+        }
+        client.set_client_metadata(client_metadata)
+
+        if token_endpoint_auth_method == 'none':
+            client.client_secret = ''
+        else:
+            client.client_secret = gen_salt(48)
+
+        db.session.add(client)
+        db.session.commit()
+        return client
+
+
+class AuthorizationCode(db.Model, OAuth2AuthorizationCodeMixin):
+    id = db.Column(db.Integer, primary_key=True)
+    user_id = db.Column(
+        db.Integer, db.ForeignKey('user.id', ondelete='CASCADE')
+    )
+    user = db.relationship('User')
+
+
+class AuthorizationCodeGrant(grants.AuthorizationCodeGrant):
+
+    def create_authorization_code(self, client, grant_user, request):
+        # you can use other method to generate this code
+        code = generate_token(48)
+        item = AuthorizationCode(
+            code=code,
+            client_id=client.client_id,
+            redirect_uri=request.redirect_uri,
+            scope=request.scope,
+            user_id=grant_user.get_user_id(),
+        )
+        db.session.add(item)
+        db.session.commit()
+        return code
+
+    def parse_authorization_code(self, code, client):
+        item = AuthorizationCode.query.filter_by(
+            code=code, client_id=client.client_id).first()
+        if item and not item.is_expired():
+            return item
+
+    def delete_authorization_code(self, authorization_code):
+        db.session.delete(authorization_code)
+        db.session.commit()
+
+    def authenticate_user(self, authorization_code):
+        return User.query.get(authorization_code.user_id)
+
+
+class PasswordGrant(grants.ResourceOwnerPasswordCredentialsGrant):
+    TOKEN_ENDPOINT_AUTH_METHODS = [
+        'client_secret_basic', 'client_secret_post'
+    ]
+
+    def authenticate_user(self, username, password):
+        user = User.query.filter_by(username=username).first()
+        if user.check_password(password):
```

Comment:
need to return something if the if is False

Reply:
Fixed by https://github.com/crs4/life_monitor/pull/21/commits/7c480e1a2b9cb66bef55bb36025153b4c31cd7e3